src/com/facebook/buck/python/PythonRunTestsStep.java

```diff
@@ -0,0 +1,160 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.python;
+
+import com.facebook.buck.shell.ShellStep;
+import com.facebook.buck.step.ExecutionContext;
+import com.facebook.buck.step.Step;
+import com.facebook.buck.test.selectors.TestDescription;
+import com.facebook.buck.test.selectors.TestSelectorList;
+import com.facebook.buck.util.HumanReadableException;
+import com.google.common.base.Function;
+import com.google.common.base.Joiner;
+import com.google.common.base.Optional;
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.regex.Pattern;
+
+public class PythonRunTestsStep implements Step {
+
+  private final Path workingDirectory;
+  private final String testName;
+  private final ImmutableList<String> commandPrefix;
+  private final Supplier<ImmutableMap<String, String>> environment;
+  private final TestSelectorList testSelectorList;
+  private final Optional<Long> testRuleTimeoutMs;
+  private final Path resultsOutputPath;
+
+  private boolean timedOut;
+
+  public PythonRunTestsStep(
+      Path workingDirectory,
+      String testName,
+      ImmutableList<String> commandPrefix,
+      Supplier<ImmutableMap<String, String>> environment,
+      TestSelectorList testSelectorList,
+      Optional<Long> testRuleTimeoutMs,
+      Path resultsOutputPath) {
+    this.workingDirectory = workingDirectory;
+    this.testName = testName;
+    this.commandPrefix = commandPrefix;
+    this.environment = environment;
+    this.testSelectorList = testSelectorList;
+    this.testRuleTimeoutMs = testRuleTimeoutMs;
+    this.resultsOutputPath = resultsOutputPath;
+
+    this.timedOut = false;
+  }
+
+  @Override
+  public int execute(ExecutionContext context) throws IOException, InterruptedException {
+    int exitCode = doExecute(context);
+    if (timedOut) {
+      throw new HumanReadableException(
+          "Following test case timed out: " + testName + ", with exitCode: " + exitCode);
+    }
+    return exitCode;
+  }
+
+  private int doExecute(ExecutionContext context) throws IOException, InterruptedException {
+    if (testSelectorList.isEmpty()) {
+      return getShellStepWithArgs("-o", resultsOutputPath.toString()).execute(context);
+    }
+
+    ShellStep listStep = getShellStepWithArgs("-l");
+    int exitCode = listStep.execute(context);
+    if (timedOut || exitCode != 0) {
+      return exitCode;
+    }
+
+    ImmutableList.Builder<String> testsToRunRe = ImmutableList.builder();
+
+    String output = listStep.getStdout();
+    for (String strTestCase : output.split("\n")) {
+      String[] testCase = strTestCase.split("#", 2);
+      if (testCase.length != 2) {
```

Comment:
I think it's valid to have test selectors `#method_name` and `class_name#`. Please add support for these and test them as well.